src/articles/2024-01-21_DDEV-for-js.md

```diff
@@ -0,0 +1,327 @@
+
+```txt
+Restarted my-project
+Your project can be reached at https://my-project.ddev.site https://api.my-project.ddev.site https://app.my-project.ddev.site
+```
+
+At this point, your new URLs will point into the webserver at the correct ports, but our apps aren't running. We could stop here and either pass commands in with `ddev exec` or tunnel in with `ddev ssh` to start the node processes inside the web container.
+
+```txt
+$ ddev ssh
+
+andy@my-project-web:/var/www/html$ cd svelte/
+andy@my-project-web:/var/www/html/svelte$ npm run dev
+```
+
+But if we run our apps this way, it's an extra step to onboard new devs with & we still have a terminal window taken over by the process. Let's boost the developer experience.
+
+## Developer Experience Improvements
+
+There are three main DX improvements we can make to make onboarding new developers as seamless as possible. First, we can utilize a process management tool like PM2 to manage our processes in the container. Second, we can make some custom commands to create easy-to-remember startup directions or to handle specific tools in the container. Third, DDEV has the ability to run commands both on the host machine and in the containers during a few key lifecycle stages.
+
+### Using PM2
+
+[PM2](https://pm2.keymetrics.io/) is an "Advanced, production process manager for Node.JS". We can use this tool to start our app without clogging up an entire terminal pane and it will also allow us to monitor the process's logs, memory usage, and some other key stats. There are a ton of options available, but we'll just use a few basics.
+
+PM2's instructions tell us to install the tool globally, so we'll handle during DDEV's build process with a custom Dockerfile in `.ddev/web-build/Dockerfile`. For more info on how DDEV handles custom Dockerfiles, see [here](https://ddev.readthedocs.io/en/stable/users/extend/customizing-images/#adding-extra-dockerfiles-for-webimage-and-dbimage)
+
+```Dockerfile
+RUN npm install -g pm2
+```
+
+We can then setup our PM2 config by adding a file `apps.config.js`. This file will name our processes, tell PM2 where to run the start command, as well as what command to run. For a full list of config options see the [relevant docs](https://pm2.keymetrics.io/docs/usage/application-declaration/)
+
+```js
+module.exports = {
+  apps : [
+    {
+      name: "Svelte",
+      cwd: "/var/www/html/svelte",
+      script: "npm run dev"
+    },
+    {
+      name: "Keystone",
+      cwd: "/var/www/html/keystone",
+      script: "npm run dev"
+    }
+  ]
+}
+```
+
+### Post-start hooks
+
+With PM2 successfully added into the container & configured for our apps, we can now utilize [DDEV's Hooks](https://ddev.readthedocs.io/en/stable/users/configuration/hooks/) to run a command on startup. In our `.ddev/config.yml` file, add the following:
+
+```yml
+hooks:
+  post-start:
+    - exec: "pm2 start apps.config.js"
```

Comment:
I think you'll like doing this better with `web_extra_daemons`, https://ddev.readthedocs.io/en/stable/users/extend/customization-extendibility/#running-extra-daemons-in-the-web-container

Reply:
What's the benefit of using `web_extra_daemons` over post-start hook?

Reply:
`web_extra_daemons` automatically starts things up. I guess it's more "integrated". But I don't see any problem with either way.